File: elixir/y2023/day1.ex
defmodule Day1 do
  def day1_1() do
    f = Path.absname("inputs/day1.txt")

    result =
      File.stream!(f)
      |> Enum.map(&String.trim/1)
      |> Enum.map(&extract_numbers/1)
      |> Enum.map(&extract_first_and_last/1)
      |> Enum.map(&String.to_integer/1)
      |> Enum.sum()

    IO.puts(result)
  end

  def extract_numbers(s) do
    s
    |> String.to_charlist()
    |> Enum.map(fn c -> c - ?0 end)
    |> Enum.filter(fn i -> i >= 0 && i <= 9 end)
    |> Enum.join("")
  end

  def extract_first_and_last(s) do
    String.first(s) <> String.last(s)
  end

  def day1_2() do
    f = Path.absname("inputs/day1.txt")

    result =
      File.stream!(f)
      |> Enum.map(&String.trim/1)
      |> Enum.map(&words_to_int/1)
      |> Enum.map(&extract_numbers/1)
      |> Enum.map(&extract_first_and_last/1)
      |> Enum.map(&String.to_integer/1)
      |> Enum.sum()

    IO.puts(result)
  end

  def words_to_int(s) do
    mappings = %{
      "oneight" => "18",
      "twone" => "21",
      "threeight" => "38",
      "fiveight" => "58",
      "eightwo" => "82",
      "eighthree" => "83",
      "nineight" => "98",
      "nine" => "9",
      "eight" => "8",
      "seven" => "7",
      "six" => "6",
      "five" => "5",
      "four" => "4",
      "three" => "3",
      "two" => "2",
      "one" => "1",
      "zero" => "0"
    }

    Enum.reduce(mappings, s, fn {key, value}, acc ->
      String.replace(acc, key, value)
    end)
  end
end

Day1.day1_1()
Day1.day1_2()

File: src/main/java/y2023/Day1.java
package y2023;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Locale;
import java.util.Scanner;

public class Day1 {
    public static void main(String[] args) throws Exception {
        Scanner scanner = IOUtils.getScanner("day1.txt");
        int sum = 0;
        while(scanner.hasNext()){
            String line = scanner.nextLine();
            line = line.toLowerCase(Locale.ENGLISH)
                    .replaceAll("oneight", "18")
                    .replaceAll("twone", "21")
                    .replaceAll("threeight", "38")
                    .replaceAll("fiveight", "58")
                    .replaceAll("sevenine", "79")
                    .replaceAll("eightwo", "82")
                    .replaceAll("eighthree", "83")
                    .replaceAll("nineight", "98")
                    .replaceAll("one", "1")
                    .replaceAll("two", "2")
                    .replaceAll("three", "3")
                    .replaceAll("four", "4")
                    .replaceAll("five", "5")
                    .replaceAll("six", "6")
                    .replaceAll("seven", "7")
                    .replaceAll("eight", "8")
                    .replaceAll("nine", "9");
                    //.replaceAll("zero", "0");
            boolean foundFirst = false;
            int lineNumber = 0;
            int secondChar = 0;
            for (char c : line.toCharArray()){
                int charNumber = c - (int) '0';
                if(charNumber >= 0 && charNumber <= 9){
                    if(!foundFirst){
                        lineNumber = charNumber * 10;
                        foundFirst = true;
                    }
                    secondChar = charNumber;
                }
            }
            lineNumber += secondChar;
            sum += lineNumber;
        }
        System.out.println(sum);
    }
}
